core/src/client/async_client/mod.rs

```diff
@@ -378,13 +398,202 @@ impl SubscriptionClientT for Client {
+				}
+				Ok(None) => (),
+				Err(err) => return Err(err),
+			}
+		}
+		// Subscription response.
+		else if let Ok(response) = serde_json::from_slice::<SubscriptionResponse<_>>(&raw) {
+			tracing::debug!("[backend]: recv subscription {:?}", response);
+			if let Err(Some(unsub)) = process_subscription_response(manager, response) {
+				let _ = stop_subscription(sender, manager, unsub).await;
+			}
+		}
+		// Subscription error response.
+		else if let Ok(response) = serde_json::from_slice::<SubscriptionError<_>>(&raw) {
+			tracing::debug!("[backend]: recv subscription closed {:?}", response);
+			let _ = process_subscription_close_response(manager, response);
+		}
+		// Incoming Notification
+		else if let Ok(notif) = serde_json::from_slice::<Notification<_>>(&raw) {
+			tracing::debug!("[backend]: recv notification {:?}", notif);
+			let _ = process_notification(manager, notif);
+		}
+		// Batch response.
+		else if let Ok(batch) = serde_json::from_slice::<Vec<Response<_>>>(&raw) {
+			tracing::debug!("[backend]: recv batch {:?}", batch);
+			if let Err(e) = process_batch_response(manager, batch) {
+				return Err(e);
+			}
+		}
+		// Error response
+		else if let Ok(err) = serde_json::from_slice::<ErrorResponse>(&raw) {
+			tracing::debug!("[backend]: recv error response {:?}", err);
+			if let Err(e) = process_error_response(manager, err) {
+				return Err(e);
+			}
+		}
+		// Unparsable response
+		else {
+			tracing::debug!(
+					"[backend]: recv unparseable message: {:?}",
+					serde_json::from_slice::<serde_json::Value>(&raw)
+				);
+			return Err(Error::Custom("Unparsable response".into()));
+		}
+		Ok(())
+	}
+
+	match message {
+		Some(Ok(ReceivedMessage::Bytes(raw))) => {
+			handle_recv_message(raw.as_ref(), manager, sender, max_notifs_per_subscription).await?;
+		}
+		Some(Ok(ReceivedMessage::Text(raw))) => {
+			handle_recv_message(raw.as_ref(), manager, sender, max_notifs_per_subscription).await?;
+		}
+		Some(Err(e)) => {
+			tracing::error!("Error: {:?} terminating client", e);
+			return Err(Error::Transport(e.into()));
+		}
+		None => {
+			tracing::error!("[backend]: WebSocket receiver dropped; terminate client");
+			return Err(Error::Custom("WebSocket receiver dropped".into()));
+		}
+	}
+
+	return Ok(());
+}
+
+/// Handle frontend messages.
+///
+/// Returns error if the main background loop should be terminated.
+async fn handle_frontend_messages<S: TransportSenderT>(
+	message: Option<FrontToBack>,
+	manager: &mut RequestManager,
+	sender: &mut S,
+	max_notifs_per_subscription: usize,
+) -> Result<(), Error> {
+	match message {
+		// User dropped the sender side of the channel.
+		// There is nothing to do just terminate.
+		None => {
+			return Err(Error::Custom("[backend]: frontend dropped; terminate client".into()));
+		}
+
+		Some(FrontToBack::Batch(batch)) => {
+			tracing::trace!("[backend]: client prepares to send batch request: {:?}", batch.raw);
+			// NOTE(niklasad1): annoying allocation.
```

Comment:
@niklasad1 DQ: Just wondering, I've seen this before and have wanted to ask, what specifically about the clone allocation is annoying here?

Reply:
we can remove that note but it's just the those are `batch ids` are already owned in this scope and so should be possible to borrow that `in theory` but we use the `HashSet entry which takes ownership of that + `RequestManager` doesn't allow borrows for simplicity.